<a href="https://colab.research.google.com/github/OviRata/Machine-Learning/blob/main/Neural_Network_Applied_on_Wine_Quality_DataSet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
%matplotlib inline
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler 
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = (20.0, 10.0)

In [3]:
!mkdir data && curl -o data/winequality-red.csv -LJO https://github.com/dianaartiom/machine_learning_with_python/raw/main/lesson05/notebooks/data/winequality-red.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 84199  100 84199    0     0   140k      0 --:--:-- --:--:-- --:--:--  140k


In [4]:
data = pd.read_csv('data/winequality-red.csv', sep=';')
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [5]:
data.quality.value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [6]:
# Collecting X and Y
y = data['quality'].values
X = data.drop(['quality'], axis=1).values

Checking for NaN values. Normalizing and scaling the data.

In [82]:
means=[];
for i in range(len(X[0]) ):
  means.append( X.mean(axis=1) );

for i in range( len(X) ):
  for j in range(len(X[i] ) ):
    if pd.isna(X[i][j]):
      X[i][j]=means[j];

mm_scaler = MinMaxScaler();
X=mm_scaler.fit_transform(X);
y=mm_scaler.fit_transform(y.reshape(-1,1));
print(X);
print(y);


[[0.24778761 0.39726027 0.         ... 0.60629921 0.13772455 0.15384615]
 [0.28318584 0.52054795 0.         ... 0.36220472 0.20958084 0.21538462]
 [0.28318584 0.43835616 0.04       ... 0.40944882 0.19161677 0.21538462]
 ...
 [0.15044248 0.26712329 0.13       ... 0.53543307 0.25149701 0.4       ]
 [0.11504425 0.35958904 0.12       ... 0.65354331 0.22754491 0.27692308]
 [0.12389381 0.13013699 0.47       ... 0.51181102 0.19760479 0.4       ]]
[[0.4]
 [0.4]
 [0.4]
 ...
 [0.6]
 [0.4]
 [0.6]]


Separating the dataset in test set and training set.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Import MLPRegressor and train the model.

In [60]:
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_squared_error

sk_model = MLPRegressor(
    random_state=42, max_iter=500, hidden_layer_sizes=(600, 600), activation='relu',
    alpha=0.0001, learning_rate_init=0.001
    ).fit(X_train, y_train)


MSE=mean_squared_error;
loss=MSE;

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1599: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Checking the accuracy of our model, by observing it's MSE, RMSE, and what portion of the total target feature interval is occupied by the error.

In [80]:
import math
y_pred_train=sk_model.predict(X_train)

error=loss(y_pred_train, y_train)
#print("The precision of the model is: ", sk_model.score(X_train, y_train)*100, "%" );
print("The MSE for training is: ", error)
print("The RMSE for training is:", math.sqrt(error))
print(f"RMSE represents {math.sqrt(error)/1.0*100}% of the total target feature interval.");

The MSE for training is:  0.012516944395377625
The RMSE for training is: 0.11187915085205834
RMSE represents 11.187915085205834% of the total target feature interval.


In [81]:
y_pred_test=sk_model.predict(X_test)

error=loss(y_pred_test, y_test)
#print("The precision of the model is: ", sk_model.accuracy_score(X_test, y_test)*100, "%" );
print("The error for testing is: ", error)
print("The RMSE for testing is:", math.sqrt(error))
print(f"RMSE represents {math.sqrt(error)/1.0*100}% of the total target feature interval.");

The error for testing is:  0.013739596357396442
The RMSE for testing is: 0.117216024320041
RMSE represents 11.721602432004099% of the total target feature interval.


The distance between adjacent unique values of the target feature is 0.2.

Meanwhile, the RMSE is around 0.117, 11.72% of the interval (0, 1).

The error is quiet great, but the result is still close to the truth.